# ğŸ”’ å³æ ¼ãƒ•ã‚¡ã‚¤ãƒ«ãƒ»ãƒ•ã‚©ãƒ«ãƒ€ä½œæˆãƒ«ãƒ¼ãƒ«

**æœ€çµ‚æ›´æ–°**: 2025-07-05  
**é©ç”¨ç¯„å›²**: å…¨æ–°è¦ãƒ•ã‚¡ã‚¤ãƒ«ãƒ»ãƒ•ã‚©ãƒ«ãƒ€ä½œæˆ  
**å„ªå…ˆåº¦**: å¿…é ˆï¼ˆproduction-grade enforcementï¼‰  
**æº–æ‹ åŸºæº–**: Enterprise Best Practices 2024-2025

## ğŸ¯ åŸºæœ¬åŸå‰‡

### 1. **Fail-Fast Policyï¼ˆé«˜é€Ÿå¤±æ•—åŸå‰‡ï¼‰**
- ä½œæˆå‰æ¤œè¨¼ã§å•é¡Œã‚’å³åº§ã«æ¤œå‡º
- ãƒ­ãƒ¼ã‚«ãƒ«ç’°å¢ƒã§å¤±æ•—ã•ã›ã€CIã¯å®‰å…¨ç¶²ã¨ã—ã¦æ©Ÿèƒ½

### 2. **Policy as Codeï¼ˆãƒãƒªã‚·ãƒ¼ã®ã‚³ãƒ¼ãƒ‰åŒ–ï¼‰**
- å…¨ãƒ«ãƒ¼ãƒ«ã¯ã‚³ãƒ¼ãƒ‰ã§å®šç¾©ãƒ»è‡ªå‹•å®Ÿè¡Œ
- æ‰‹å‹•ãƒã‚§ãƒƒã‚¯ã¯ä¸€åˆ‡ä¸è¦

### 3. **Cross-Platform Safetyï¼ˆã‚¯ãƒ­ã‚¹ãƒ—ãƒ©ãƒƒãƒˆãƒ•ã‚©ãƒ¼ãƒ å®‰å…¨æ€§ï¼‰**
- Windows, macOS, Linux, ã‚¯ãƒ©ã‚¦ãƒ‰ã‚¹ãƒˆãƒ¬ãƒ¼ã‚¸å¯¾å¿œ
- å¤§æ–‡å­—å°æ–‡å­—ã®é•ã„ã«ã‚ˆã‚‹è¡çªé˜²æ­¢

## ğŸ“‹ ä½œæˆå‰æ¤œè¨¼ãƒã‚§ãƒƒã‚¯ãƒªã‚¹ãƒˆ

### **Phase 1: å‘½åæ¤œè¨¼**
```yaml
naming_validation:
  # æ–‡å­—ã‚»ãƒƒãƒˆæ¤œè¨¼
  allowed_chars: "^[a-z0-9]+(-[a-z0-9]+)*$"
  
  # é•·ã•åˆ¶é™
  max_length:
    file_name: 50
    folder_name: 40
    full_path: 240
  
  # ç¦æ­¢ãƒ‘ã‚¿ãƒ¼ãƒ³
  forbidden_patterns:
    - "^[0-9]"          # æ•°å­—é–‹å§‹ç¦æ­¢
    - "--"              # é€£ç¶šãƒã‚¤ãƒ•ãƒ³ç¦æ­¢
    - "-$"              # ãƒã‚¤ãƒ•ãƒ³çµ‚äº†ç¦æ­¢
    - "^-"              # ãƒã‚¤ãƒ•ãƒ³é–‹å§‹ç¦æ­¢
  
  # OSäºˆç´„èª
  reserved_words:
    windows: ["con", "prn", "aux", "nul", "com1-9", "lpt1-9"]
    unix: [".", "..", "~"]
    
  # æ‹¡å¼µå­ãƒ«ãƒ¼ãƒ«
  extension_rules:
    multiple_dots: false  # file.test.js ç¦æ­¢
    max_extension_length: 10
```

### **Phase 2: æ§‹é€ æ¤œè¨¼**
```yaml
structure_validation:
  # éšå±¤æ·±åº¦ãƒã‚§ãƒƒã‚¯
  max_depth: 5
  
  # ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªé…ç½®ãƒ«ãƒ¼ãƒ«
  placement_rules:
    scripts: 
      allowed: [".sh", ".py"]
      forbidden: [".md", ".txt", ".json"]
    docs:
      allowed: [".md", ".txt", ".rst"]
      forbidden: [".sh", ".py", ".js"]
    config:
      allowed: [".json", ".yml", ".yaml", ".toml", ".env"]
      forbidden: [".sh", ".py", ".md"]
    src:
      allowed: [".py", ".js", ".ts", ".jsx", ".tsx"]
      forbidden: [".sh", ".md"]
```

### **Phase 3: ã‚»ã‚­ãƒ¥ãƒªãƒ†ã‚£æ¤œè¨¼**
```yaml
security_validation:
  # æ©Ÿå¯†æƒ…å ±ãƒ‘ã‚¿ãƒ¼ãƒ³æ¤œå‡º
  sensitive_patterns:
    filenames: ["secret", "key", "password", "token", "credential"]
    content_scan: true
    
  # PII/ãƒ‡ãƒ¼ã‚¿åˆ†é¡
  data_classification:
    required_tags: ["-public", "-internal", "-confidential"]
    tag_position: "suffix"
    
  # ã‚¢ã‚¯ã‚»ã‚¹åˆ¶å¾¡
  access_control:
    production_paths: ["prod/", "production/"]
    restricted_to: ["svc-prod-deploy", "admin-role"]
```

### **Phase 4: ãƒ¡ã‚¿ãƒ‡ãƒ¼ã‚¿è¦ä»¶**
```yaml
metadata_requirements:
  # å¿…é ˆãƒ¡ã‚¿ãƒ‡ãƒ¼ã‚¿
  required_fields:
    - owner         # æ‰€æœ‰è€…/ãƒãƒ¼ãƒ 
    - purpose       # ç›®çš„
    - created_date  # ä½œæˆæ—¥æ™‚ï¼ˆUTCï¼‰
    - expires       # æœ‰åŠ¹æœŸé™ï¼ˆä¸€æ™‚ãƒ•ã‚¡ã‚¤ãƒ«ã®å ´åˆï¼‰
    
  # ã‚¿ã‚¤ãƒ ã‚¹ã‚¿ãƒ³ãƒ—å½¢å¼
  timestamp_format: "yyyymmddThhmmssZ"  # 20250705T164500Z
  
  # ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆè¦ä»¶
  documentation:
    readme_required: true  # æ–°è¦ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒª
    changelog_required: true  # å¤‰æ›´å±¥æ­´
```

## ğŸ¤– è‡ªå‹•å®Ÿè¡Œãƒ¡ã‚«ãƒ‹ã‚ºãƒ 

### **1. Pre-Creation Hook**
```bash
#!/bin/bash
# .githooks/pre-create-validate.sh

validate_name() {
    local name="$1"
    local type="$2"  # file or folder
    
    # æ–‡å­—ã‚»ãƒƒãƒˆæ¤œè¨¼
    if ! [[ "$name" =~ ^[a-z0-9]+(-[a-z0-9]+)*$ ]]; then
        echo "âŒ Invalid $type name: $name"
        echo "   Must use only lowercase letters, numbers, and hyphens"
        return 1
    fi
    
    # é•·ã•æ¤œè¨¼
    if [ "$type" = "file" ] && [ ${#name} -gt 50 ]; then
        echo "âŒ File name too long: ${#name} chars (max: 50)"
        return 1
    fi
    
    # äºˆç´„èªãƒã‚§ãƒƒã‚¯
    if is_reserved_word "$name"; then
        echo "âŒ Reserved word detected: $name"
        return 1
    fi
    
    return 0
}

validate_placement() {
    local file_path="$1"
    local dir=$(dirname "$file_path")
    local ext="${file_path##*.}"
    
    # ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªåˆ¥é…ç½®ãƒ«ãƒ¼ãƒ«é©ç”¨
    case "$dir" in
        */scripts/*)
            if [[ "$ext" =~ ^(md|txt|json)$ ]]; then
                echo "âŒ Document files not allowed in scripts/"
                return 1
            fi
            ;;
        */docs/*)
            if [[ "$ext" =~ ^(sh|py|js)$ ]]; then
                echo "âŒ Executable files not allowed in docs/"
                return 1
            fi
            ;;
    esac
    
    return 0
}

# ãƒ¡ã‚¤ãƒ³æ¤œè¨¼ãƒ•ãƒ­ãƒ¼
main() {
    local target_path="$1"
    local name=$(basename "$target_path")
    local type="file"
    
    [ -d "$target_path" ] && type="folder"
    
    # å…¨æ¤œè¨¼å®Ÿè¡Œ
    validate_name "$name" "$type" || exit 1
    validate_placement "$target_path" || exit 1
    validate_security "$target_path" || exit 1
    create_metadata "$target_path" || exit 1
    
    echo "âœ… All validations passed for: $target_path"
}

main "$@"
```

### **2. IDEçµ±åˆï¼ˆVSCodeè¨­å®šä¾‹ï¼‰**
```json
{
  "files.participants": [
    {
      "id": "file-creation-validator",
      "onWillCreateFiles": {
        "command": "validateFileCreation",
        "arguments": ["${file}"]
      }
    }
  ],
  "validateFileCreation.rules": {
    "namingPattern": "^[a-z0-9]+(-[a-z0-9]+)*$",
    "maxLength": 50,
    "autoFix": true
  }
}
```

### **3. CI/CDãƒ‘ã‚¤ãƒ—ãƒ©ã‚¤ãƒ³çµ±åˆ**
```yaml
# .github/workflows/file-validation.yml
name: File Structure Validation

on:
  pull_request:
    types: [opened, synchronize]

jobs:
  validate:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      
      - name: Validate File Names
        run: |
          python3 scripts/validate-file-structure.py --strict
          
      - name: Check Placement Rules
        run: |
          python3 scripts/check-placement-rules.py
          
      - name: Security Scan
        run: |
          gitleaks detect --source . --verbose
          
      - name: Generate Compliance Report
        if: always()
        run: |
          python3 scripts/generate-compliance-report.py > compliance.sarif
          
      - uses: github/codeql-action/upload-sarif@v2
        with:
          sarif_file: compliance.sarif
```

## ğŸ“Š å®Ÿè£…å„ªå…ˆé †ä½

### **Phase 1: å³åº§å®Ÿè£…ï¼ˆDay 1ï¼‰**
1. å‘½åè¦å‰‡æ¤œè¨¼
2. æ–‡å­—ã‚»ãƒƒãƒˆåˆ¶é™
3. é•·ã•åˆ¶é™
4. Pre-commitãƒ•ãƒƒã‚¯

### **Phase 2: çŸ­æœŸå®Ÿè£…ï¼ˆWeek 1ï¼‰**
1. ãƒ‡ã‚£ãƒ¬ã‚¯ãƒˆãƒªé…ç½®ãƒ«ãƒ¼ãƒ«
2. äºˆç´„èªãƒ–ãƒ­ãƒƒã‚¯
3. IDEçµ±åˆ
4. åŸºæœ¬ã‚»ã‚­ãƒ¥ãƒªãƒ†ã‚£ã‚¹ã‚­ãƒ£ãƒ³

### **Phase 3: ä¸­æœŸå®Ÿè£…ï¼ˆMonth 1ï¼‰**
1. ãƒ¡ã‚¿ãƒ‡ãƒ¼ã‚¿è‡ªå‹•ç”Ÿæˆ
2. ã‚¢ã‚¯ã‚»ã‚¹åˆ¶å¾¡çµ±åˆ
3. å®Œå…¨CI/CDçµ±åˆ
4. ã‚³ãƒ³ãƒ—ãƒ©ã‚¤ã‚¢ãƒ³ã‚¹ãƒ¬ãƒãƒ¼ãƒˆ

## ğŸ” æ¤œè¨¼ä¾‹

### âœ… æ­£ã—ã„ä¾‹
```bash
# ãƒ•ã‚¡ã‚¤ãƒ«ä½œæˆ
src/user-management/auth-service.py
docs/api/user-guide.md
config/environments/production.yml
scripts/deploy-app.sh

# ãƒ•ã‚©ãƒ«ãƒ€ä½œæˆ
src/payment-processing/
tests/integration/
docs/architecture/
```

### âŒ é–“é•ã£ãŸä¾‹
```bash
# é•åä¾‹ã¨ç†ç”±
src/UserManagement/       # å¤§æ–‡å­—ç¦æ­¢
scripts/README.md         # ãƒ‰ã‚­ãƒ¥ãƒ¡ãƒ³ãƒˆã¯ docs/ ã¸
config/prod_settings.yml  # ã‚¢ãƒ³ãƒ€ãƒ¼ã‚¹ã‚³ã‚¢ç¦æ­¢
src/api/test.sh          # å®Ÿè¡Œã‚¹ã‚¯ãƒªãƒ—ãƒˆã¯ scripts/ ã¸
docs/api-key.txt         # æ©Ÿå¯†æƒ…å ±ãƒ‘ã‚¿ãƒ¼ãƒ³æ¤œå‡º
```

## ğŸš€ è‡ªå‹•ä¿®æ­£æ©Ÿèƒ½

```python
# scripts/auto-fix-naming.py
import re
from pathlib import Path

class FileNameAutoFixer:
    def fix_name(self, name: str) -> str:
        """ä¸æ­£ãªåå‰ã‚’è‡ªå‹•ä¿®æ­£"""
        # å¤§æ–‡å­—ã‚’å°æ–‡å­—ã«
        name = name.lower()
        
        # ç‰¹æ®Šæ–‡å­—ã‚’ãƒã‚¤ãƒ•ãƒ³ã«
        name = re.sub(r'[^a-z0-9\-\.]', '-', name)
        
        # é€£ç¶šãƒã‚¤ãƒ•ãƒ³ã‚’å˜ä¸€ã«
        name = re.sub(r'-+', '-', name)
        
        # å…ˆé ­æœ«å°¾ã®ãƒã‚¤ãƒ•ãƒ³ã‚’å‰Šé™¤
        name = name.strip('-')
        
        # é•·ã•åˆ¶é™
        if len(name) > 50:
            name = name[:47] + '...'
            
        return name
```

## ğŸ“ˆ ã‚³ãƒ³ãƒ—ãƒ©ã‚¤ã‚¢ãƒ³ã‚¹æŒ‡æ¨™

### **è¿½è·¡ãƒ¡ãƒˆãƒªã‚¯ã‚¹**
- Pre-commitå¤±æ•—ç‡ï¼ˆç›®æ¨™: < 5%ï¼‰
- å‘½åè¦å‰‡é•åæ•°/é€±ï¼ˆç›®æ¨™: 0ï¼‰
- è‡ªå‹•ä¿®æ­£æˆåŠŸç‡ï¼ˆç›®æ¨™: > 90%ï¼‰
- ãƒãƒªã‚·ãƒ¼æº–æ‹ ç‡ï¼ˆç›®æ¨™: 100%ï¼‰

### **ç›£æŸ»ãƒ­ã‚°**
```json
{
  "timestamp": "2025-07-05T16:45:00Z",
  "action": "file_created",
  "path": "src/auth/login-service.py",
  "validations": {
    "naming": "passed",
    "placement": "passed",
    "security": "passed",
    "metadata": "generated"
  },
  "user": "dev-user-123",
  "auto_fixed": false
}
```

---

**ğŸ”’ ã“ã®ãƒ«ãƒ¼ãƒ«ã«ã‚ˆã‚Šã€ã‚¨ãƒ³ã‚¿ãƒ¼ãƒ—ãƒ©ã‚¤ã‚ºã‚°ãƒ¬ãƒ¼ãƒ‰ã®ãƒ•ã‚¡ã‚¤ãƒ«ä½œæˆæ¨™æº–ã‚’ç¢ºä¿ã—ã¾ã™**

**âš¡ å…¨æ¤œè¨¼ã¯è‡ªå‹•åŒ–ã•ã‚Œã€é–‹ç™ºè€…ã®ç”Ÿç”£æ€§ã‚’æãªã†ã“ã¨ãªãå“è³ªã‚’ä¿è¨¼ã—ã¾ã™**